### ISSUE TITLE
Bug: [no-inferrable-types] False positive with default parameter in closure passed to generic function

### ISSUE THREAD

**Issue by 6XGate** [typescript-eslint/typescript-eslint#9130]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.3&fileType=.tsx&code=PTAEHkGlQWlAxArgOwMYBcCWB7ZoAG6ApgM7r6iYkEAUAbgIYA2iRAXKMogLYBGRAJwCUoALwA%2BUHWyYAJhQbJZoWdlKds6UAKIBHRJh0E2XPoIoALTMnQA6AFAgCybDGsAzQQIa8mRGOgAngAOpBRU2gxURMqo2AI6GEyBDkQAHsHxWu4oGDh4xGT0zKwcpvwCYqAArCIA3vagoHHIJNh%2BtkzYAObFLERC9gC%2B9o5gULAIuVi4BIXoAEzh1Ph9pZw8FSISUjLytqAActigXvHUEd7RsgA0zfGJ6MmpGVmgOWgzBaSLa0RVtVADSaLTaHS6vUY-UGIzGU0%2B%2BVA6AsDC0AHdMEwmKAAERo7zBHGgRSBd7TRGKZTcKI2GmUdDUIKhBwfPKzfEMYIAHgAggJutR0sQlNQUABrFxo5AAbQAuuIaO5kBwaLY1Qx%2BSQOHyBdtJNI5PVGtoiOhEAI8Kr1ZrtZq9UDjSDcGCiJ0ejQAFIAZXAh1sZAE1m6mHcgRoGt1gyaTR0Zot72QVtsEZIUdAI1hTmOWgmcHwUNYFDR2EQTGU-AI4slyHC7mM5XMoCsNko1G4VBIQYcTnwLjcyE8CR8fgCITCrci10oaAeRCSgTuyP%2BTP%2BGKxcJcWgrHi8MRemQEWlBWnmAGYqhzgjQ-mVNoIAfbgfdWu1XRC-jDBnDsxBoHmC0QyygCgshEO41h7qAAAqFjLmOoBogw1CbtOg46LI%2B5vMeSI-AALBeBLXgBD5iJIT6gq%2BbqQiUAzDIMQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDNFpoAQwBGSWkVKoMkftA7RI4MAF8QyoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
// OK - Function `test` is `(value: number) => void` and does not require `:number` hint.
// `no-inferrable-types` is raised correctly.
export function test(value: number = 5) {
  console.log(value)
}

// OK - Function `test2` is `(value: number) => void`. No errors is raised, correctly.
export function test2(value = 5) {
  console.log(value)
}

// Function that will "wrap" any function and maintain its type.
function wrap<Args extends unknown[]>(fn: (...args: Args) => void) {
  return (...args: Args) => {
    console.log(JSON.stringify(args))
    return fn(...args)
  }
}

// Not OK - `value` would be `unknown` if `:number` hint is missing.
// `no-inferrable-types` is raised incorrectly, the type will
// not be inferred.
export const test3 = wrap((value: number = 5) => {
  console.log(value)
})

// Not OK - `value` is undefined. The type was not inferred.
export const test4 = wrap((value = 5) => {
  console.log(value)
})
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-inferrable-types": "error"
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Default parameters of any function passed to another **generic** function must be typed without raising `@typescript-eslint/no-inferrable-types`.

### Actual Result

Default parameters of any function passed to another **generic** function that are properly typed raises `@typescript-eslint/no-inferrable-types`.

### Additional Info

When passing a function to another generic function that infers argument types, default parameters of the passed function must be typed.
